File: 2023/internal/fifteen/day15.go
package fifteen

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/josuemolinamorales/aoc-2023/utils/iterators"
)

const testInput = `rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7`

func RunDayFifteen() {
	input, err := os.ReadFile("./input/day15.txt")
	if err != nil {
		panic("Failed to read day 15 file")
	}
	fmt.Println("Part one", partOne(string(input)))
	fmt.Println("Part two", partTwo(string(input)))
}

func partOne(input string) int {
	sum := 0
	for _, line := range strings.Split(input, ",") {
		sum += hash(line)
	}

	return sum
}

func hash(input string) int {
	current := 0
	for _, char := range input {
		current += int(char)
		current *= 17
		current %= 256
	}

	return current
}

type Value struct {
	Key   string
	Value int
}

type Box struct {
	Values []Value
}

func partTwo(input string) int {
	hm := make([]Box, 256)
	for _, line := range strings.Split(input, ",") {
		key, op, value := "", "", 0
		for i, char := range line {
			if char == '-' {
				op = "remove"
				break
			}
			if char == '=' {
				n, _ := strconv.Atoi(line[i+1:])
				value = n
				op = "add"
				break
			}
			key += string(char)
		}
		idx := hash(key)
		matchKey := func(v Value) bool {
			return v.Key == key
		}
		switch op {
		case "add":
			// Find the value to change
			if bi, item := iterators.Find(hm[idx].Values, matchKey); bi != -1 {
				// Key found in box, replace it
				item.Value = value
				hm[idx].Values[bi] = *item
			} else {
				// If not found, add it to the end of the box
				hm[idx].Values = append(hm[idx].Values, Value{Key: key, Value: value})
			}
		case "remove":
			// Find the item to remove
			if i, _ := iterators.Find(hm[idx].Values, matchKey); i != -1 {
				hm[idx].Values = append(hm[idx].Values[:i], hm[idx].Values[i+1:]...)
			}
		}
	}
	sum := 0
	for i, box := range hm {
		for j, value := range box.Values {
			sum += (i + 1) * (j + 1) * value.Value
		}
	}

	return sum
}

File: 2015/solutions/day15.py

def parse(input):
    ingredients = []
    for line in input.splitlines():
        line = line.split()
        ingredients.append({
            "capacity": int(line[2][:-1]),
            "durability": int(line[4][:-1]),
            "flavor": int(line[6][:-1]),
            "texture": int(line[8][:-1]),
            "calories": int(line[10]),
        })
    return ingredients


def part_one(input):
    ingredients = parse(input)

    max_score = 0
    for a in range(101):
        for b in range(101-a):
            for c in range(101-a-b):
                d = 100-a-b-c
                capacity = max(0, a*ingredients[0]["capacity"] + b*ingredients[1]
                               ["capacity"] + c*ingredients[2]["capacity"] + d*ingredients[3]["capacity"])
                durability = max(0, a*ingredients[0]["durability"] + b*ingredients[1]
                                 ["durability"] + c*ingredients[2]["durability"] + d*ingredients[3]["durability"])
                flavor = max(0, a*ingredients[0]["flavor"] + b*ingredients[1]
                             ["flavor"] + c*ingredients[2]["flavor"] + d*ingredients[3]["flavor"])
                texture = max(0, a*ingredients[0]["texture"] + b*ingredients[1]
                              ["texture"] + c*ingredients[2]["texture"] + d*ingredients[3]["texture"])
                score = capacity * durability * flavor * texture
                max_score = max(max_score, score)

    return max_score


def part_two(input):
    ingredients = parse(input)

    max_score = 0
    for a in range(101):
        for b in range(101-a):
            for c in range(101-a-b):
                d = 100-a-b-c
                calories = a*ingredients[0]["calories"] + b*ingredients[1]["calories"] + \
                    c*ingredients[2]["calories"] + d*ingredients[3]["calories"]
                if calories == 500:
                    capacity = max(0, a*ingredients[0]["capacity"] + b*ingredients[1]
                                   ["capacity"] + c*ingredients[2]["capacity"] + d*ingredients[3]["capacity"])
                    durability = max(0, a*ingredients[0]["durability"] + b*ingredients[1]
                                     ["durability"] + c*ingredients[2]["durability"] + d*ingredients[3]["durability"])
                    flavor = max(0, a*ingredients[0]["flavor"] + b*ingredients[1]
                                 ["flavor"] + c*ingredients[2]["flavor"] + d*ingredients[3]["flavor"])
                    texture = max(0, a*ingredients[0]["texture"] + b*ingredients[1]
                                  ["texture"] + c*ingredients[2]["texture"] + d*ingredients[3]["texture"])
                    score = capacity * durability * flavor * texture
                    max_score = max(max_score, score)

    return max_score


def run_day_fifteen():
    with open("./input/day15.txt", "r") as f:
        input = f.read()
    print("Day 15 part one:", part_one(input))
    print("Day 15 part two:", part_two(input))
